Improved use of containers

marcosmarp opened the pull request:
Solves #16

marcosmarp commented:
Just noticed that i forgot about the range-for thing! I will upload it in a couple hours.

nahueespinosa commented on main.cpp:
```diff
@@ -182,29 +185,35 @@ void RegisterNewUser() {
   PrintTab(1);
   std::cout << "Choose an option[1-5]: ";
-  std::cin >> registered_users_data[amount_of_registered_users-1].security_question_choice;
+  std::cin >> security_question_choice_temporal;
```
Why don't you use your already temporal object to store this? 

```cpp
std::cin >> new_user_temporal.security_question_choice;
```

Same with other fields.

nahueespinosa commented on main.cpp:
```diff
@@ -93,8 +93,8 @@ void RegisterNewUser() {
 
-    amount_of_registered_users--; // Since the registration process was started but never finished, we have to decrease the amount of registered users
-    registered_users_data.resize(amount_of_registered_users);
+    //amount_of_registered_users--; // Since the registration process was started but never finished, we have to decrease the amount of registered users
```
Remove commented out code. Let it go haha.

marcosmarp replied:
LOL. I didn't deleted cause the last time i deleted a line, all of the code below it appeared as "new" or "modified" when i created the PR, just because they were moved a line

nahueespinosa commented on main.cpp:
```diff
@@ -68,17 +67,18 @@ void PrintTab(int numberOfTabs) {
-  short dot_position, at_sign_position;
-  std::string  email_temporal, username_temporal;
+  short dot_position, at_sign_position, security_question_choice_temporal;
+  std::string  email_temporal, username_temporal, password_temporal, security_question_answer_temporal;
```
You won't need any of these variables if you use `new_user_temporal` to store the temporal values. See comment below.

marcosmarp replied:
Will delete them :)

nahueespinosa commented on main.cpp:
```diff
@@ -250,7 +259,7 @@ void LogIn() { // The other one of the main functions, goes through the log in p
   unsigned long username_position;
 
-  for(username_position = 0 ; username_position < amount_of_registered_users ; username_position++) {
+  for(username_position = 0 ; username_position < registered_users_data.size() ; username_position++) {
```
This is a nice spot for a **range-for** :smiley:

nahueespinosa commented on main.cpp:
```diff
@@ -355,7 +364,7 @@ void RecoverPasswordQuestionMethod(unsigned long username_position) { // This is
   int i;
 
-  for(i = 0 ; i < amount_of_registered_users ; i++) {
+  for(i = 0 ; i < registered_users_data.size() ; i++) {
```
This is asking for a **range-for**

nahueespinosa commented on main.cpp:
```diff
@@ -365,24 +374,24 @@ bool CheckIfEmailAlreadyRegistered (std::string entered_email) { // This functio
   int i;
 
-  for(i = 0 ; i < amount_of_registered_users ; i++) {
+  for(i = 0 ; i < registered_users_data.size() ; i++) {
```
Have I mentioned **range-for** before? (?

marcosmarp replied:
Lmao, i'm going to dream with **range-for** xD. Will add it to the code!

nahueespinosa reviewed (changes requested):
I left you some **range-for**s, I mean, comments hahaha

marcosmarp commented:
Hey, i can't see to fully understand the range-for. For example:

```
long array[5];
for (auto& i : array) {
  i++;
}
```
Will this increase in one each position of `array`? If that's so, why are we increasing `i` instead of `array`? Thanks!

marcosmarp commented:
Got it! A question:
`elem = 0.1f` what do the `f` stands for? Why don't just `elem = 0.1`?

nahueespinosa commented:
It's a [literal suffix](https://www.learncpp.com/cpp-tutorial/literals/) to create a float literal.

Things you want to pay attention when working with floats are [implicit conversions](https://www.cplusplus.com/doc/tutorial/typecasting/) and loss of precision.

This should be a good read: https://www.learncpp.com/cpp-tutorial/floating-point-numbers/

marcosmarp commented:
Okay! That was interesting, i think I got it.

Is standard practice to always specify the literal suffix when working with floats? I'm asking cause in this cause, if i didn't misunderstood, since `elem` was declared as a `float`, there shouldn't be any implicit conversion, right?

nahueespinosa commented:
This initializes a float variable with a double literal so it is an implicit conversion.
```cpp
float elem = 0.1;
```
This creates the float representation of 0.1 and initializes a float variable with it (no conversion).
```cpp
float elem = 0.1f;
```
The standard practice would be to use `double` unless you have a good reason to use `float`. So:
```cpp
double elem = 0.1;
```

marcosmarp commented:
Got it! Will implement the changes. Thanks!

marcosmarp commented:
Added the **range-for** (they really make the code much nicer; looking forward for more tips like that one) and deleted all of the temporal variables in the registration process, using directly the temporal UserData one

nahueespinosa reviewed (changes requested):
Hey, left some comments!

marcosmarp commented on main.cpp:
```diff
@@ -362,22 +345,17 @@ void RecoverPasswordQuestionMethod(unsigned long username_position) { // This is
-
-  for(i = 0 ; i < registered_users_data.size() ; i++) {
-    if (registered_users_data[i].email == entered_email) return 1;
+  for(auto& i : registered_users_data) {
+    if (i.email == entered_email) return 1;
```
That looks better!  Will add it.

marcosmarp commented on main.cpp:
```diff
@@ -259,8 +239,11 @@ void LogIn() {
 
-  for(username_position = 0 ; username_position < registered_users_data.size() ; username_position++) {
-    if (entered_username == registered_users_data[username_position].username) return username_position;
+  for(auto& i : registered_users_data) {
```
Sorry, i think i'm missing something.
This is what i got right now:
```
std::vector<UserData>::iterator FindUserByName(std::string entered_username) {
  std::vector<UserData>::iterator i;

  for(i = registered_users_data.begin() ; i != registered_users_data.end() ; i++) {
    if (entered_username == registered_users_data[std::distance(registered_users_data.begin(), i)].username) return i;
  }
  return i;
}
```
Is there something i'm doing wrong? I don't seem to find the advantage of the iterator, in fact, having to calculate the distance between the beginning position of the vector with the current one to find the current position of it seems like a hassle; i'm sure there's something i didn't catch

nahueespinosa replied:
You can use the iterator to access the data, it is a well behaved pointer.

```cpp
std::vector<UserData>::iterator FindUserByName(std::string entered_username) {
  for (auto it = registered_users_data.begin(); it != registered_users_data.end(); ++it) {
    if (entered_username == it->username)
      break;
  }
  return it;
}
```

Notice how I define the iterator inside the `for` statement. This is a nice feature compared to C language, keep your variable definitions close to where they're being used.

It doesn't matter much here, but in general is good practice to prefer [`pre-increment` over `post-increment`](https://en.cppreference.com/w/cpp/language/operator_incdec) operators. I'll let you google why :wink:.

Only one `return` for function is a good guideline, unless it obfuscates the code.

marcosmarp commented:
Okay! I think i got it know. Tell me what you think

By the way, regarding this:
>Notice how I define the iterator inside the for statement. This is a nice feature compared to C language, keep your variable definitions close to where they're being used.

I tried to do it that way but the compiler keep throwing an error in the `return it` instance": it was not defined" so i declared it the traditional way.

nahueespinosa reviewed (changes requested):
Hey! Almost there. Let's use [pass-by-reference](https://www.w3schools.com/cpp/cpp_function_reference.asp) parameters and merge this so we can move on to classes and OOP.

marcosmarp commented:
Implemented the corrections! Tell me what you think :)

nahueespinosa reviewed (approved):
Great work!

marcosmarp commented:
Thanks!! Merging them
